## **BAMS 506 Final Project: Diet Planning Problem**

 **Teammates:**
 - Novar Qin 92408798
 - Evie Xu 75527671
 - Kasundra Betinol 31712417; 

### **Part A - Algebraically Formulation**


##### Define Variables:
$$\begin{array}{rll}
 \text {Xi: Number of servings of food item i} \\
\end{array}
$$

##### Objective Function:
$$
\begin{array}{rll}
 \text{Minimize:} \sum_{i=1}^{30} c_i \cdot X_i \\
\end{array}
$$

$$
\begin{array}{rll}
\text{Where $c_i$ is the cost per serving of food item 𝑖, and $𝑥_𝑖$ is the number of servings of food item 𝑖. }
\end{array}
$$

##### Constraints
1.	Calories constraint:
$$
\begin{array}{rll}
    \sum_{i=1}^{n}Calories_i \cdot X_i \ge\ 1800 \ & \text{(Lower Limit for Calories)}\\
    \sum_{i=1}^{n}Calories_i \cdot X_i \leq\ 2400 \ & \text{(Upper Limit for Calories)}\\
    \ Calories_i \leq\ 0.3 \cdot \sum_{i=1}^{n}Calories_i \ & \text{(Balanced Diet)}\\
\end{array}
$$

2.	Fat constraint:
$$
\begin{array}{rll}
    \sum_{i=1}^{n}Fat_i \cdot X_i \ge\ 60 \ & \text{(Lower Limit for Fat)}\\
    \sum_{i=1}^{n}Fat_i \cdot X_i \leq\ 95 \ & \text{(Upper Limit for Fat)}\\
\end{array}
$$

3.	Sodium constraint:
$$
\begin{array}{rll}
    \sum_{i=1}^{n}Sodium_i \cdot X_i \ge\ 1200 \ & \text{(Lower Limit for Sodium)}\\
    \sum_{i=1}^{n}Sodium_i \cdot X_i \leq\ 2200 \ & \text{(Upper Limit for Sodium)}\\
\end{array}
$$

4.	Carbs constraint:
$$
\begin{array}{rll}
    \sum_{i=1}^{n}Carbs_i \cdot X_i \ge\ 240 \ & \text{(Lower Limit for Carbs)}\\
    \sum_{i=1}^{n}Carbs_i \cdot X_i \leq\ 400 \ & \text{(Upper Limit for Carbs)}\\
\end{array}
$$

5.	Fiber constraint:
$$
\begin{array}{rll}
    \sum_{i=1}^{n}Fiber_i \cdot X_i \ge\ 30 \ & \text{(Lower Limit for Fiber)}\\
    \sum_{i=1}^{n}Fiber_i \cdot X_i \leq\ 35 \ & \text{(Upper Limit for Fiber)}\\
\end{array}
$$

6.	Protein constraint:
$$
\begin{array}{rll}
    \sum_{i=1}^{n}Protein_i \cdot X_i \ge\ 40 \ & \text{(Lower Limit for Protein)}\\
    \sum_{i=1}^{n}Protein_i \cdot X_i \leq\ 55 \ & \text{(Upper Limit for Protein)}\\
    \ Protein_i \leq\ 0.3 \cdot \sum_{i=1}^{n}Protein_i \ & \text{(Balanced Diet)}\\
\end{array}
$$

7.	Vitamin A constraint:
$$
\begin{array}{rll}
    \sum_{i=1}^{n}VitaminA_i \cdot X_i \ge\ 2000 \ & \text{(Lower Limit for VitaminA)}\\
    \sum_{i=1}^{n}VitaminA_i \cdot X_i \leq\ 6000 \ & \text{(Upper Limit for VitaminA)}\\
\end{array}
$$

8.	Vitamin C constraint:
$$
\begin{array}{rll}
    \sum_{i=1}^{n}VitaminC_i \cdot X_i \ge\ 45 \ & \text{(Lower Limit for VitaminC)}\\
    \sum_{i=1}^{n}VitaminC_i \cdot X_i \leq\ 1200 \ & \text{(Upper Limit for VitaminC)}\\
\end{array}
$$

9.	Calcium constraint:
$$
\begin{array}{rll}
    \sum_{i=1}^{n}Calcium_i \cdot X_i \ge\ 1300 \ & \text{(Lower Limit for Calcium)}\\
    \sum_{i=1}^{n}Calcium_i \cdot X_i \leq\ 3000 \ & \text{(Upper Limit for Calcium)}\\
\end{array}
$$

10.	Iron constraint:
$$
\begin{array}{rll}
    \sum_{i=1}^{n}Iron_i \cdot X_i \ge\ 1300 \ & \text{(Lower Limit for Iron)}\\
    \sum_{i=1}^{n}Iron_i \cdot X_i \leq\ 3000 \ & \text{(Upper Limit for Iron)}\\
\end{array}
$$


### **Part B - Gurobi Model** 

##### **Run Gurobi Model**

In [16]:
import gurobipy as gp
from gurobipy import GRB

# Create a new model
model = gp.Model("NutritionalOptimization")

# Data: Coefficients for food items (cost, calories, fat, sodium, etc.)
n = 30  # Number of food items

# Replace these arrays with the actual data from your table
cost = [0.16, 0.07, 0.18, 0.02, 0.53, 0.06, 0.31, 0.84, 0.78, 0.27, 0.24, 0.15, 0.32, 0.49, 0.15, 0.16, 0.05, 0.06, 0.23, 0.13, 0.08, 0.11, 0.15, 0.27, 0.82, 0.39, 0.08, 0.17, 0.81, 0.69]      # Cost per serving of each food item
calories = [73.8, 23.7, 72.2, 2.6, 20, 171.5, 88.2, 277.4, 358.2, 25.8, 81.4, 104.9, 15.1, 46.4, 61.6, 78, 65, 65, 121.2, 85.5, 74.5, 99.6, 56.4, 141.8, 145.1, 100.8, 103, 98.7, 710.8, 115.6]  # Calories per serving
fat = [0.8, 0.1, 0.6, 0, 0.1, 0.2, 5.5, 10.8, 12.3, 0.4, 0.5, 0.5, 0.1, 0.3, 0.2, 0.5, 1, 1, 4.7, 0.4, 5, 7.3, 4.3, 12.8, 2.3, 0.1, 0, 0.5, 72.2, 2.1]       # Fat per serving
sodium = [68.2, 19.2, 2.5, 1.8, 1.5, 15.2, 8.1, 125.6, 1237.1, 11.1, 0, 1.1, 0.5, 3.8, 0, 151.4, 134.5, 132.5, 121.8, 126.2, 140, 168, 248.9, 461.7, 2.3, 4.5, 0.2, 0.7, 38.4, 333.2]    # Sodium per serving
carbs = [13.6, 5.6, 17.1, 0.4, 4.8, 39.9, 2.2, 0, 58.3, 5.7, 21, 26.7, 4.1, 11.3, 15.4, 15.1, 12.4, 11.8, 11.7, 11.9, 0.6, 1.3, 0.3, 0.8, 25.3, 20.9, 0.8, 19.8, 0, 0]     # Carbohydrates per serving
fiber = [8.5, 1.6, 2, 0.3, 1.3, 3.2, 1.4, 0, 11.6, 1.4, 3.7, 2.7, 0.2, 2.6, 3.1, 0.6, 1.3, 1.1, 0, 0, 0, 0, 0, 0, 4, 1.3, 22.3, 0.9, 0, 0]     # Fiber per serving
protein = [8, 0.6, 2.5, 0.2, 0.7, 3.7, 9.4, 42.2, 8.2, 1, 0.3, 1.2, 0.2, 0.8, 1.2, 3, 2.2, 2.3, 8.1, 8.4, 6.2, 6.7, 3.9, 5.4, 6.1, 3.4, 0.3, 3.3, 13.8, 22.7]   # Protein per serving
vitaminA = [5867.4, 15471, 106.6, 66, 467.7, 0, 98.6, 77.4, 3055.2, 766.3, 73.1, 92.3, 24, 133, 268.6, 0, 0, 0, 500.2, 499.8, 316, 409.2, 0, 0, 37.4, 0, 2.1, 0, 14.7, 68]  # Vitamin A per serving
vitaminC = [160.2, 5.1, 5.2, 0.8, 66.1, 15.6, 0.1, 0, 27.9, 23.5, 7.9, 10.4, 1, 74.5, 69.7, 0, 0, 0, 2.3, 2.4, 0, 0.1, 0, 10.8, 0, 0, 0, 0, 0, 0]  # Vitamin C per serving
calcium = [159, 14.9, 3.3, 3.8, 6.7, 22.7, 121.8, 21.9, 80.2, 6.2, 9.7, 6.8, 3.4, 19.8, 52.4, 21, 10.8, 26.2, 296.7, 302.3, 24.5, 42.6, 23.8, 9, 18.7, 7.2, 0, 4.9, 59.9, 3.4]   # Calcium per serving
iron = [2.3, 0.3, 0.3, 0.1, 0.3, 4.3, 6.2, 1.8, 2.3, 0.6, 0.2, 0.4, 0.1, 0.3, 0.1, 1, 0.7, 0.8, 0.1, 0.1, 0.7, 0.7, 0.4, 0.6, 1.6, 0.3, 7.9, 1, 0.4, 0.5]      # Iron per serving

# Food items dictionary
food_items = {
    1: "Broccoli",
    2: "Carrots, Raw",
    3: "Corn",
    4: "Lettuce, Iceberg,Raw",
    5: "Peppers, Sweet, Raw",
    6: "Potatoes, Baked",
    7: "Tofu",
    8: "Roasted Chicken",
    9: "Spaghetti W/ Sauce",
    10: "Tomato,Red,Ripe,Raw",
    11: "Apple, Raw, w/Skin",
    12: "Banana",
    13: "Grapes",
    14: "Kiwifruit, Raw, Fresh",
    15: "Oranges",
    16: "Bagels",
    17: "Wheat Bread",
    18: "White Bread",
    19: "2pcnt Lowfat Milk",
    20: "Skim Milk",
    21: "Poached Eggs",
    22: "Scrambled Eggs",
    23: "Turkey",
    24: "Beef",
    25: "Oatmeal",
    26: "Couscous",
    27: "White Rice",
    28: "Macaroni, cooked",
    29: "Pork",
    30: "White Tuna in Water"
}

# Create decision variables (number of servings of each food item)
x = model.addVars(n, vtype=GRB.CONTINUOUS, name="x")

# Objective: Minimize cost
model.setObjective(gp.quicksum(cost[i] * x[i] for i in range(n)), GRB.MINIMIZE)

# Constraints:

# Calories constraint
model.addConstr(gp.quicksum(calories[i] * x[i] for i in range(n)) >= 1800, "calories_min")
model.addConstr(gp.quicksum(calories[i] * x[i] for i in range(n)) <= 2400, "calories_max")
for i in range(n):
    model.addConstr(calories[i] * x[i] <= 0.3 * gp.quicksum(calories[j] * x[j] for j in range(n)))

# Fat constraint
model.addConstr(gp.quicksum(fat[i] * x[i] for i in range(n)) >= 60, "fat_min")
model.addConstr(gp.quicksum(fat[i] * x[i] for i in range(n)) <= 95, "fat_max")

# Sodium constraint
model.addConstr(gp.quicksum(sodium[i] * x[i] for i in range(n)) >= 1200, "sodium_min")
model.addConstr(gp.quicksum(sodium[i] * x[i] for i in range(n)) <= 2200, "sodium_max")

# Carbohydrates constraint
model.addConstr(gp.quicksum(carbs[i] * x[i] for i in range(n)) >= 240, "carbs_min")
model.addConstr(gp.quicksum(carbs[i] * x[i] for i in range(n)) <= 400, "carbs_max")

# Fiber constraint
model.addConstr(gp.quicksum(fiber[i] * x[i] for i in range(n)) >= 30, "fiber_min")
model.addConstr(gp.quicksum(fiber[i] * x[i] for i in range(n)) <= 35, "fiber_max")

# Protein constraint
model.addConstr(gp.quicksum(protein[i] * x[i] for i in range(n)) >= 40, "protein_min")
model.addConstr(gp.quicksum(protein[i] * x[i] for i in range(n)) <= 55, "protein_max")
for i in range(n):
    model.addConstr(protein[i] * x[i] <= 0.3 * gp.quicksum(protein[j] * x[j] for j in range(n)))

# Vitamin A constraint
model.addConstr(gp.quicksum(vitaminA[i] * x[i] for i in range(n)) >= 2000, "vitaminA_min")
model.addConstr(gp.quicksum(vitaminA[i] * x[i] for i in range(n)) <= 6000, "vitaminA_max")

# Vitamin C constraint
model.addConstr(gp.quicksum(vitaminC[i] * x[i] for i in range(n)) >= 45, "vitaminC_min")
model.addConstr(gp.quicksum(vitaminC[i] * x[i] for i in range(n)) <= 1200, "vitaminC_max")

# Calcium constraint
model.addConstr(gp.quicksum(calcium[i] * x[i] for i in range(n)) >= 1300, "calcium_min")
model.addConstr(gp.quicksum(calcium[i] * x[i] for i in range(n)) <= 3000, "calcium_max")

# Iron constraint
model.addConstr(gp.quicksum(iron[i] * x[i] for i in range(n)) >= 8, "iron_min")
model.addConstr(gp.quicksum(iron[i] * x[i] for i in range(n)) <= 40, "iron_max")

# Solve the model
model.optimize()

# Print results
if model.status == GRB.OPTIMAL:
    for i in range(n):
        if x[i].x > 0:
            print(f"{food_items[i+1]}: {x[i].x} servings")
    print(f"Total cost per serving ($): {model.objVal}")

# Print top 3 contributors for each nutrition
    nutrients = {
        "Calories": calories,
        "Fat": fat,
        "Sodium": sodium,
        "Carbohydrates": carbs,
        "Fiber": fiber,
        "Protein": protein,
        "Vitamin A": vitaminA,
        "Vitamin C": vitaminC,
        "Calcium": calcium,
        "Iron": iron
    }

    for nutrient, values in nutrients.items():
        total_contribution = sum(values[i] * x[i].x for i in range(n) if x[i].x > 0)
        contributions = [(food_items[i + 1], (values[i] * x[i].x) / total_contribution * 100) for i in range(n) if x[i].x > 0]
        contributions.sort(key=lambda x: x[1], reverse=True)
        top_3 = contributions[:3]
        print(f"Top 3 contributors for {nutrient}:")
        for food, percentage in top_3:
            print(f"  {food}: {percentage:.2f}%")
        print("")

else:
    print("No optimal solution found.")


Gurobi Optimizer version 11.0.3 build v11.0.3rc0 (mac64[arm] - Darwin 23.6.0 23G93)

CPU model: Apple M1 Pro
Thread count: 10 physical cores, 10 logical processors, using up to 10 threads

Optimize a model with 80 rows, 30 columns and 2326 nonzeros
Model fingerprint: 0xb87c4c0b
Coefficient statistics:
  Matrix range     [6e-02, 2e+04]
  Objective range  [2e-02, 8e-01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [8e+00, 6e+03]
Presolve removed 10 rows and 0 columns
Presolve time: 0.00s
Presolved: 70 rows, 40 columns, 2073 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   2.294375e+02   0.000000e+00      0s
      20    2.4310277e+00   0.000000e+00   0.000000e+00      0s

Solved in 20 iterations and 0.00 seconds (0.00 work units)
Optimal objective  2.431027707e+00
Potatoes, Baked: 3.14868804664723 servings
Spaghetti W/ Sauce: 0.6109381088589423 servings
Banana: 0.09414308492014076 servings
Oranges: 4.540457768798042 servings
2pc

##### **Result & Discussion**


This optimization result represents a meal plan designed to minimize the total cost while certain nutritional requirements. The optimal objective value of **$2.43** per serving reflects the cheapest cost of the selected combination of food items that satisfy the nutritional goals.

The meal plan includes the following portions per serving:

- Potatoes, Baked: 3.15 servings
- Spaghetti with Sauce: 0.61 servings
- Banana: 0.09 servings
- Oranges: 4.54 servings
- 2% Lowfat Milk: 2.04 servings
- Skim Milk: 1.00 serving
- White Rice: 0.16 servings
- Pork: 0.57 servings

### **Part C - Model Improvement**

##### **Problems for Original Model**

The resulted combination of food in Part B is not ideal for a realistic daily diets for several reasons:
* **Fractional Servings**: Since we set the orginal model decision variables as continuous, the serving size for many food items returns out to be fractions (e.g. 0.09 servings of banana), which is not practical for food consumption in real life.
* **Food Variety and Balance** : a nutrious diet should be balanced across different food groups. Nutrients need to come from different sources of food, rather than rely on a single source of food. From the original model, 81% of Vitamin C comes from Oranges, and 79% of Iron comes from Potatos. This over-reliance on a limited set of foods makes the diet repetitive and overlooks the need for balance across different food groups, which is crucial for maintaining long-term health for students.

##### **Updates to the New Model**

Based on above discussion, we adjusted our model with four key changes:

- **Added upper and lower bounds to make portion sizes more reasonable [0.5, 3].**
- **Divided food into different groups:**
    - Vegetables: Broccoli, Carrots, Corn, Lettuce, Peppers, Potatoes, Tomato
    - Fruits: Apple, Banana, Grapes, Kiwi, Oranges
    - Protein: Tofu, Roasted Chicken, Poached Eggs, Turkey, Beef, Pork, White Tuna
    - Grains: Spaghetti, Bagels, Wheat Bread, White Bread, Oatmeal, Couscous, White Rice, Macaroni
    - Dairy: 2% Milk, Skim Milk
- **Ensured that at least one food from each category is selected to maintain a balanced diet.**
- **Ensured that no single food item contributes more than 50% of the total value needed for each nutrient.**

While constructing this model, we also attempted to set the ratio and absolute amount of food from each category in alignment with Canada's Food Guidelines. However, the resulting costs were too high for an already economically burdensome plan. In the end, we decided to forgo this idea.

##### **Run New Gurobi Model**

In [15]:
import gurobipy as gp
from gurobipy import GRB
import math

# Create a new model
model = gp.Model("NutritionalOptimization")

# Data: Coefficients for food items (cost, calories, fat, sodium, etc.)
n = 30  # Number of food items

# Replace these arrays with the actual data from your table
cost = [0.16, 0.07, 0.18, 0.02, 0.53, 0.06, 0.31, 0.84, 0.78, 0.27, 0.24, 0.15, 0.32, 0.49, 0.15, 0.16, 0.05, 0.06, 0.23, 0.13, 0.08, 0.11, 0.15, 0.27, 0.82, 0.39, 0.08, 0.17, 0.81, 0.69]      # Cost per serving of each food item
calories = [73.8, 23.7, 72.2, 2.6, 20, 171.5, 88.2, 277.4, 358.2, 25.8, 81.4, 104.9, 15.1, 46.4, 61.6, 78, 65, 65, 121.2, 85.5, 74.5, 99.6, 56.4, 141.8, 145.1, 100.8, 103, 98.7, 710.8, 115.6]  # Calories per serving
fat = [0.8, 0.1, 0.6, 0, 0.1, 0.2, 5.5, 10.8, 12.3, 0.4, 0.5, 0.5, 0.1, 0.3, 0.2, 0.5, 1, 1, 4.7, 0.4, 5, 7.3, 4.3, 12.8, 2.3, 0.1, 0, 0.5, 72.2, 2.1]       # Fat per serving
sodium = [68.2, 19.2, 2.5, 1.8, 1.5, 15.2, 8.1, 125.6, 1237.1, 11.1, 0, 1.1, 0.5, 3.8, 0, 151.4, 134.5, 132.5, 121.8, 126.2, 140, 168, 248.9, 461.7, 2.3, 4.5, 0.2, 0.7, 38.4, 333.2]    # Sodium per serving
carbs = [13.6, 5.6, 17.1, 0.4, 4.8, 39.9, 2.2, 0, 58.3, 5.7, 21, 26.7, 4.1, 11.3, 15.4, 15.1, 12.4, 11.8, 11.7, 11.9, 0.6, 1.3, 0.3, 0.8, 25.3, 20.9, 0.8, 19.8, 0, 0]     # Carbohydrates per serving
fiber = [8.5, 1.6, 2, 0.3, 1.3, 3.2, 1.4, 0, 11.6, 1.4, 3.7, 2.7, 0.2, 2.6, 3.1, 0.6, 1.3, 1.1, 0, 0, 0, 0, 0, 0, 4, 1.3, 22.3, 0.9, 0, 0]     # Fiber per serving
protein = [8, 0.6, 2.5, 0.2, 0.7, 3.7, 9.4, 42.2, 8.2, 1, 0.3, 1.2, 0.2, 0.8, 1.2, 3, 2.2, 2.3, 8.1, 8.4, 6.2, 6.7, 3.9, 5.4, 6.1, 3.4, 0.3, 3.3, 13.8, 22.7]   # Protein per serving
vitaminA = [5867.4, 15471, 106.6, 66, 467.7, 0, 98.6, 77.4, 3055.2, 766.3, 73.1, 92.3, 24, 133, 268.6, 0, 0, 0, 500.2, 499.8, 316, 409.2, 0, 0, 37.4, 0, 2.1, 0, 14.7, 68]  # Vitamin A per serving
vitaminC = [160.2, 5.1, 5.2, 0.8, 66.1, 15.6, 0.1, 0, 27.9, 23.5, 7.9, 10.4, 1, 74.5, 69.7, 0, 0, 0, 2.3, 2.4, 0, 0.1, 0, 10.8, 0, 0, 0, 0, 0, 0]  # Vitamin C per serving
calcium = [159, 14.9, 3.3, 3.8, 6.7, 22.7, 121.8, 21.9, 80.2, 6.2, 9.7, 6.8, 3.4, 19.8, 52.4, 21, 10.8, 26.2, 296.7, 302.3, 24.5, 42.6, 23.8, 9, 18.7, 7.2, 0, 4.9, 59.9, 3.4]   # Calcium per serving
iron = [2.3, 0.3, 0.3, 0.1, 0.3, 4.3, 6.2, 1.8, 2.3, 0.6, 0.2, 0.4, 0.1, 0.3, 0.1, 1, 0.7, 0.8, 0.1, 0.1, 0.7, 0.7, 0.4, 0.6, 1.6, 0.3, 7.9, 1, 0.4, 0.5]      # Iron per serving

# Food items dictionary
food_items = {
    1: "Broccoli",
    2: "Carrots, Raw",
    3: "Corn",
    4: "Lettuce, Iceberg,Raw",
    5: "Peppers, Sweet, Raw",
    6: "Potatoes, Baked",
    7: "Tofu",
    8: "Roasted Chicken",
    9: "Spaghetti W/ Sauce",
    10: "Tomato,Red,Ripe,Raw",
    11: "Apple, Raw, w/Skin",
    12: "Banana",
    13: "Grapes",
    14: "Kiwifruit, Raw, Fresh",
    15: "Oranges",
    16: "Bagels",
    17: "Wheat Bread",
    18: "White Bread",
    19: "2pcnt Lowfat Milk",
    20: "Skim Milk",
    21: "Poached Eggs",
    22: "Scrambled Eggs",
    23: "Turkey",
    24: "Beef",
    25: "Oatmeal",
    26: "Couscous",
    27: "White Rice",
    28: "Macaroni, cooked",
    29: "Pork",
    30: "White Tuna in Water"
}

# Create decision variables (number of servings of each food item)
x = model.addVars(n, vtype=GRB.SEMICONT, lb=0.5, ub=3, name="x")

# Objective: Minimize cost
model.setObjective(gp.quicksum(cost[i] * x[i] for i in range(n)), GRB.MINIMIZE)

# Mapping of food items to groups (ensure each index corresponds correctly to your items list)
groups = {
   'Vegetables': [1, 2, 3, 4, 5, 6, 10], # Broccoli, Carrits, Corn, Lettuce, Peppers, Potatoes, Tomato
    'Fruits': [11, 12, 13, 14, 15], # Apple, Banana, Grapes, Kiwi, Oranges
    'Protein': [7, 8, 21, 22, 23, 24, 29, 30], #Tofu, Roasted Chicken, Poached Eggs, Turkey, Beef, Pork, White Tuna
    'Grains': [9, 16, 17, 18, 25, 26, 27, 28], #Spaghetti, Bagels, Wheat Bread, White Bread, Oatmeal, Couscous, White Rice, Macaroni
    'Dairy': [19, 20] # 2% Milk, Skim Milk  
}

# Add constraints to ensure at least one item from each group is chosen
for group in groups:
    model.addConstr(gp.quicksum(x[i-1] for i in groups[group]) >= 1, f"at_least_one_{group}")

# Add constraint to ensure no single food item contributes more than half of each nutrient
for i in range(n):
    model.addConstr(calories[i] * x[i] <= 0.5 * gp.quicksum(calories[j] * x[j] for j in range(n)), f"calories_diverse_{i}")
    model.addConstr(fat[i] * x[i] <= 0.5 * gp.quicksum(fat[j] * x[j] for j in range(n)), f"fat_diverse_{i}")
    model.addConstr(sodium[i] * x[i] <= 0.5 * gp.quicksum(sodium[j] * x[j] for j in range(n)), f"sodium_diverse_{i}")
    model.addConstr(carbs[i] * x[i] <= 0.5 * gp.quicksum(carbs[j] * x[j] for j in range(n)), f"carbs_diverse_{i}")
    model.addConstr(fiber[i] * x[i] <= 0.5 * gp.quicksum(fiber[j] * x[j] for j in range(n)), f"fiber_diverse_{i}")
    model.addConstr(protein[i] * x[i] <= 0.5 * gp.quicksum(protein[j] * x[j] for j in range(n)), f"protein_diverse_{i}")
    model.addConstr(vitaminA[i] * x[i] <= 0.5 * gp.quicksum(vitaminA[j] * x[j] for j in range(n)), f"vitaminA_diverse_{i}")
    model.addConstr(vitaminC[i] * x[i] <= 0.5 * gp.quicksum(vitaminC[j] * x[j] for j in range(n)), f"vitaminC_diverse_{i}")
    model.addConstr(calcium[i] * x[i] <= 0.5 * gp.quicksum(calcium[j] * x[j] for j in range(n)), f"calcium_diverse_{i}")
    model.addConstr(iron[i] * x[i] <= 0.5 * gp.quicksum(iron[j] * x[j] for j in range(n)), f"iron_diverse_{i}")
    
# Calories constraint
model.addConstr(gp.quicksum(calories[i] * x[i] for i in range(n)) >= 1800, "calories_min")
model.addConstr(gp.quicksum(calories[i] * x[i] for i in range(n)) <= 2400, "calories_max")
for i in range(n):
    model.addConstr(calories[i] * x[i] <= 0.3 * gp.quicksum(calories[j] * x[j] for j in range(n)))

# Fat constraint
model.addConstr(gp.quicksum(fat[i] * x[i] for i in range(n)) >= 60, "fat_min")
model.addConstr(gp.quicksum(fat[i] * x[i] for i in range(n)) <= 95, "fat_max")

# Sodium constraint
model.addConstr(gp.quicksum(sodium[i] * x[i] for i in range(n)) >= 1200, "sodium_min")
model.addConstr(gp.quicksum(sodium[i] * x[i] for i in range(n)) <= 2200, "sodium_max")

# Carbohydrates constraint
model.addConstr(gp.quicksum(carbs[i] * x[i] for i in range(n)) >= 240, "carbs_min")
model.addConstr(gp.quicksum(carbs[i] * x[i] for i in range(n)) <= 400, "carbs_max")

# Fiber constraint
model.addConstr(gp.quicksum(fiber[i] * x[i] for i in range(n)) >= 30, "fiber_min")
model.addConstr(gp.quicksum(fiber[i] * x[i] for i in range(n)) <= 35, "fiber_max")

# Protein constraint
model.addConstr(gp.quicksum(protein[i] * x[i] for i in range(n)) >= 40, "protein_min")
model.addConstr(gp.quicksum(protein[i] * x[i] for i in range(n)) <= 55, "protein_max")
for i in range(n):
    model.addConstr(protein[i] * x[i] <= 0.3 * gp.quicksum(protein[j] * x[j] for j in range(n)))

# Vitamin A constraint
model.addConstr(gp.quicksum(vitaminA[i] * x[i] for i in range(n)) >= 2000, "vitaminA_min")
model.addConstr(gp.quicksum(vitaminA[i] * x[i] for i in range(n)) <= 6000, "vitaminA_max")

# Vitamin C constraint
model.addConstr(gp.quicksum(vitaminC[i] * x[i] for i in range(n)) >= 45, "vitaminC_min")
model.addConstr(gp.quicksum(vitaminC[i] * x[i] for i in range(n)) <= 1200, "vitaminC_max")

# Calcium constraint
model.addConstr(gp.quicksum(calcium[i] * x[i] for i in range(n)) >= 1300, "calcium_min")
model.addConstr(gp.quicksum(calcium[i] * x[i] for i in range(n)) <= 3000, "calcium_max")

# Iron constraint
model.addConstr(gp.quicksum(iron[i] * x[i] for i in range(n)) >= 8, "iron_min")
model.addConstr(gp.quicksum(iron[i] * x[i] for i in range(n)) <= 40, "iron_max")

# Solve the model
model.optimize()

# Print results
if model.status == GRB.OPTIMAL:
    print(f"Total cost per serving ($): {model.objVal}\n")
    
    for group in groups:
        print(f"{group} selections:")
        for i in groups[group]:
            if x[i - 1].x > 0:  # Since 'i' starts from 1, adjust index for zero-based x
                print(f"  {food_items[i]}: {x[i - 1].x:.2f} servings")
        print("")  # Print a newline between groups for better readability

    # Print top 3 contributors for each nutrition
    nutrients = {
        "Calories": calories,
        "Fat": fat,
        "Sodium": sodium,
        "Carbohydrates": carbs,
        "Fiber": fiber,
        "Protein": protein,
        "Vitamin A": vitaminA,
        "Vitamin C": vitaminC,
        "Calcium": calcium,
        "Iron": iron
    }

    for nutrient, values in nutrients.items():
        total_contribution = sum(values[i] * x[i].x for i in range(n) if x[i].x > 0)
        contributions = [(food_items[i + 1], (values[i] * x[i].x) / total_contribution * 100) for i in range(n) if x[i].x > 0]
        contributions.sort(key=lambda x: x[1], reverse=True)
        top_3 = contributions[:3]
        print(f"Top 3 contributors for {nutrient}:")
        for food, percentage in top_3:
            print(f"  {food}: {percentage:.2f}%")
        print("")

else:
    print("No optimal solution found.")

Gurobi Optimizer version 11.0.3 build v11.0.3rc0 (mac64[arm] - Darwin 23.6.0 23G93)

CPU model: Apple M1 Pro
Thread count: 10 physical cores, 10 logical processors, using up to 10 threads

Optimize a model with 385 rows, 30 columns and 10246 nonzeros
Model fingerprint: 0xd00d0ee2
Variable types: 0 continuous, 0 integer (0 binary)
Semi-Variable types: 30 continuous, 0 integer
Coefficient statistics:
  Matrix range     [5e-02, 2e+04]
  Objective range  [2e-02, 8e-01]
  Bounds range     [5e-01, 3e+00]
  RHS range        [1e+00, 6e+03]
Presolve removed 50 rows and 1 columns
Presolve time: 0.00s
Presolved: 393 rows, 58 columns, 8887 nonzeros
Variable types: 29 continuous, 29 integer (29 binary)

Root relaxation: objective 2.893624e+00, 33 iterations, 0.00 seconds (0.01 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

     0     0    2.89362    0    3          -    2.89362      

##### **Result & Discussion**

The optimal solution yields a higher cost of **$4.85** per serving. The resulting diet plan includes a balanced selection of foods from each food group, ensuring a variety of nutrient sources. 

The selected foods are as follows:

Vegetables selections:
- Lettuce, Iceberg,Raw: 0.50 servings
- Potatoes, Baked: 0.86 servings

Fruits selections:
- Apple, Raw, w/Skin: 3.00 servings
- Grapes: 2.98 servings
- Kiwifruit, Raw, Fresh: 1.79 servings
- Oranges: 3.00 servings

Protein selections:
- Beef: 1.27 servings
- Pork: 0.50 servings

Grains selections:
- Spaghetti W/ Sauce: 0.51 servings
- White Bread: 0.50 servings

Dairy selections:
- 2pcnt Lowfat Milk: 2.04 servings
- Skim Milk: 1.15 servings

### **Part D - Weekly Plan**

##### **Run Gurobi Model for Weekly Meal Plan**

In [2]:
import gurobipy as gp
from gurobipy import GRB
import math

# Create a new model
model = gp.Model("NutritionalOptimization")

# Data: Coefficients for food items (cost, calories, fat, sodium, etc.)
n = 30  # Number of food items

# Replace these arrays with the actual data from your table
cost = [0.16, 0.07, 0.18, 0.02, 0.53, 0.06, 0.31, 0.84, 0.78, 0.27, 0.24, 0.15, 0.32, 0.49, 0.15, 0.16, 0.05, 0.06, 0.23, 0.13, 0.08, 0.11, 0.15, 0.27, 0.82, 0.39, 0.08, 0.17, 0.81, 0.69]      # Cost per serving of each food item
calories = [73.8, 23.7, 72.2, 2.6, 20, 171.5, 88.2, 277.4, 358.2, 25.8, 81.4, 104.9, 15.1, 46.4, 61.6, 78, 65, 65, 121.2, 85.5, 74.5, 99.6, 56.4, 141.8, 145.1, 100.8, 103, 98.7, 710.8, 115.6]  # Calories per serving
fat = [0.8, 0.1, 0.6, 0, 0.1, 0.2, 5.5, 10.8, 12.3, 0.4, 0.5, 0.5, 0.1, 0.3, 0.2, 0.5, 1, 1, 4.7, 0.4, 5, 7.3, 4.3, 12.8, 2.3, 0.1, 0, 0.5, 72.2, 2.1]       # Fat per serving
sodium = [68.2, 19.2, 2.5, 1.8, 1.5, 15.2, 8.1, 125.6, 1237.1, 11.1, 0, 1.1, 0.5, 3.8, 0, 151.4, 134.5, 132.5, 121.8, 126.2, 140, 168, 248.9, 461.7, 2.3, 4.5, 0.2, 0.7, 38.4, 333.2]    # Sodium per serving
carbs = [13.6, 5.6, 17.1, 0.4, 4.8, 39.9, 2.2, 0, 58.3, 5.7, 21, 26.7, 4.1, 11.3, 15.4, 15.1, 12.4, 11.8, 11.7, 11.9, 0.6, 1.3, 0.3, 0.8, 25.3, 20.9, 0.8, 19.8, 0, 0]     # Carbohydrates per serving
fiber = [8.5, 1.6, 2, 0.3, 1.3, 3.2, 1.4, 0, 11.6, 1.4, 3.7, 2.7, 0.2, 2.6, 3.1, 0.6, 1.3, 1.1, 0, 0, 0, 0, 0, 0, 4, 1.3, 22.3, 0.9, 0, 0]     # Fiber per serving
protein = [8, 0.6, 2.5, 0.2, 0.7, 3.7, 9.4, 42.2, 8.2, 1, 0.3, 1.2, 0.2, 0.8, 1.2, 3, 2.2, 2.3, 8.1, 8.4, 6.2, 6.7, 3.9, 5.4, 6.1, 3.4, 0.3, 3.3, 13.8, 22.7]   # Protein per serving
vitaminA = [5867.4, 15471, 106.6, 66, 467.7, 0, 98.6, 77.4, 3055.2, 766.3, 73.1, 92.3, 24, 133, 268.6, 0, 0, 0, 500.2, 499.8, 316, 409.2, 0, 0, 37.4, 0, 2.1, 0, 14.7, 68]  # Vitamin A per serving
vitaminC = [160.2, 5.1, 5.2, 0.8, 66.1, 15.6, 0.1, 0, 27.9, 23.5, 7.9, 10.4, 1, 74.5, 69.7, 0, 0, 0, 2.3, 2.4, 0, 0.1, 0, 10.8, 0, 0, 0, 0, 0, 0]  # Vitamin C per serving
calcium = [159, 14.9, 3.3, 3.8, 6.7, 22.7, 121.8, 21.9, 80.2, 6.2, 9.7, 6.8, 3.4, 19.8, 52.4, 21, 10.8, 26.2, 296.7, 302.3, 24.5, 42.6, 23.8, 9, 18.7, 7.2, 0, 4.9, 59.9, 3.4]   # Calcium per serving
iron = [2.3, 0.3, 0.3, 0.1, 0.3, 4.3, 6.2, 1.8, 2.3, 0.6, 0.2, 0.4, 0.1, 0.3, 0.1, 1, 0.7, 0.8, 0.1, 0.1, 0.7, 0.7, 0.4, 0.6, 1.6, 0.3, 7.9, 1, 0.4, 0.5]      # Iron per serving

# Food items dictionary
food_items = {
    1: "Broccoli",
    2: "Carrots, Raw",
    3: "Corn",
    4: "Lettuce, Iceberg,Raw",
    5: "Peppers, Sweet, Raw",
    6: "Potatoes, Baked",
    7: "Tofu",
    8: "Roasted Chicken",
    9: "Spaghetti W/ Sauce",
    10: "Tomato,Red,Ripe,Raw",
    11: "Apple, Raw, w/Skin",
    12: "Banana",
    13: "Grapes",
    14: "Kiwifruit, Raw, Fresh",
    15: "Oranges",
    16: "Bagels",
    17: "Wheat Bread",
    18: "White Bread",
    19: "2pcnt Lowfat Milk",
    20: "Skim Milk",
    21: "Poached Eggs",
    22: "Scrambled Eggs",
    23: "Turkey",
    24: "Beef",
    25: "Oatmeal",
    26: "Couscous",
    27: "White Rice",
    28: "Macaroni, cooked",
    29: "Pork",
    30: "White Tuna in Water"
}

# Mapping of food items to groups (ensure each index corresponds correctly to your items list)
group_index = {
    'Vegetables': [1, 2, 3, 4, 5, 6, 10],
    'Fruits': [11, 12, 13, 14, 15],
    'Protein': [7, 8, 21, 22, 23, 24, 29, 30],
    'Grains': [9, 16, 17, 18, 25, 26, 27, 28],
    'Dairy': [19, 20]
}

# Track the number of times each food item is used across all solutions
food_usage = [0] * n

# Create a model to find multiple optimal solutions
model = gp.Model("NutritionalOptimization")

# Create decision variables (number of servings of each food item)
x = model.addVars(n, vtype=GRB.SEMICONT, lb=0.5, ub=3, name="x")

# Create binary variables to indicate if a food item is selected or not
y = model.addVars(n, vtype=GRB.BINARY, name="y")

# Link binary variables to serving variables using regular constraints
for i in range(n):
    model.addConstr(x[i] <= 3 * y[i], f"link_ub_{i}")
    model.addConstr(x[i] >= 0.5 * y[i], f"link_lb_{i}")

# Objective: Minimize cost 
model.setObjective(gp.quicksum(cost[i] * x[i] for i in range(n)), GRB.MINIMIZE)

# Add constraints to ensure at least one item from each group is chosen
for group, items in group_index.items():
    model.addConstr(gp.quicksum(x[i - 1] for i in items) >= 1, f"at_least_one_{group}")

# Nutritional constraints (same as in the original model)
model.addConstr(gp.quicksum(calories[i] * x[i] for i in range(n)) >= 1800, "calories_min")
model.addConstr(gp.quicksum(calories[i] * x[i] for i in range(n)) <= 2400, "calories_max")
model.addConstr(gp.quicksum(fat[i] * x[i] for i in range(n)) >= 60, "fat_min")
model.addConstr(gp.quicksum(fat[i] * x[i] for i in range(n)) <= 95, "fat_max")
model.addConstr(gp.quicksum(sodium[i] * x[i] for i in range(n)) >= 1200, "sodium_min")
model.addConstr(gp.quicksum(sodium[i] * x[i] for i in range(n)) <= 2200, "sodium_max")
model.addConstr(gp.quicksum(carbs[i] * x[i] for i in range(n)) >= 240, "carbs_min")
model.addConstr(gp.quicksum(carbs[i] * x[i] for i in range(n)) <= 400, "carbs_max")
model.addConstr(gp.quicksum(fiber[i] * x[i] for i in range(n)) >= 30, "fiber_min")
model.addConstr(gp.quicksum(fiber[i] * x[i] for i in range(n)) <= 35, "fiber_max")
model.addConstr(gp.quicksum(protein[i] * x[i] for i in range(n)) >= 40, "protein_min")
model.addConstr(gp.quicksum(protein[i] * x[i] for i in range(n)) <= 55, "protein_max")
model.addConstr(gp.quicksum(vitaminA[i] * x[i] for i in range(n)) >= 2000, "vitaminA_min")
model.addConstr(gp.quicksum(vitaminA[i] * x[i] for i in range(n)) <= 6000, "vitaminA_max")
model.addConstr(gp.quicksum(vitaminC[i] * x[i] for i in range(n)) >= 45, "vitaminC_min")
model.addConstr(gp.quicksum(vitaminC[i] * x[i] for i in range(n)) <= 1200, "vitaminC_max")
model.addConstr(gp.quicksum(calcium[i] * x[i] for i in range(n)) >= 1300, "calcium_min")
model.addConstr(gp.quicksum(calcium[i] * x[i] for i in range(n)) <= 3000, "calcium_max")
model.addConstr(gp.quicksum(iron[i] * x[i] for i in range(n)) >= 8, "iron_min")
model.addConstr(gp.quicksum(iron[i] * x[i] for i in range(n)) <= 40, "iron_max")

# Find and store the top 7 solutions in descending order of optimality
meal_plans = []
previous_solutions = []
for solution_number in range(7):
    model.optimize()

    if model.status == GRB.OPTIMAL:
        solution = {
            "total_cost": model.objVal,
            "selections": {}
        }
        for group, items in group_index.items():
            solution["selections"][group] = []
            for i in items:
                if x[i - 1].x > 0:
                    solution["selections"][group].append((food_items[i], x[i - 1].x))
                    food_usage[i - 1] += 1  # Update food usage count
        meal_plans.append(solution)

        # Store the current solution's binary variables
        current_solution = [y[i].x for i in range(n)]
        previous_solutions.append(current_solution)

        # Add a constraint to ensure each subsequent plan has at least 9 different items
        diff_count = model.addVar(vtype=GRB.INTEGER, name=f"diff_count_{solution_number}")
        model.addConstr(diff_count == gp.quicksum(y[i] + previous_solutions[-1][i] - 2 * y[i] * previous_solutions[-1][i] for i in range(n)), f"diff_count_calc_{solution_number}")
        model.addConstr(diff_count >= 8, f"different_items_{solution_number}")
    else:
        break

# Print results for all solutions
for idx, plan in enumerate(meal_plans):
    print(f"\nMeal Plan {idx + 1}:")
    print(f"Total cost per serving ($): {plan['total_cost']}")
    for group, selections in plan["selections"].items():
        print(f"{group} selections:")
        for item, servings in selections:
            print(f"  {item}: {servings} servings")


Gurobi Optimizer version 11.0.3 build v11.0.3rc0 (mac64[arm] - Darwin 21.6.0 21G72)

CPU model: Apple M2
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 85 rows, 60 columns and 676 nonzeros
Model fingerprint: 0xccdeacca
Variable types: 0 continuous, 30 integer (30 binary)
Semi-Variable types: 30 continuous, 0 integer
Coefficient statistics:
  Matrix range     [1e-01, 2e+04]
  Objective range  [2e-02, 8e-01]
  Bounds range     [5e-01, 3e+00]
  RHS range        [1e+00, 6e+03]
Presolve removed 4 rows and 2 columns
Presolve time: 0.00s
Presolved: 139 rows, 87 columns, 726 nonzeros
Variable types: 29 continuous, 58 integer (58 binary)

Root relaxation: objective 2.556965e+00, 37 iterations, 0.00 seconds (0.00 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

     0     0    2.55696    0    5          -    2.55696      -     -   

#### **Results and Discussion**

 In part C, we updated our model to ensure reasonable portions, different food groups and made sure no single food item contributed more than 50% of the total value needed for each nutrient, all while minimizing total cost per serving. In our following adjustments, our focus was to expand variety within these food groups for an entire week, as kids don't like to eat the exact thing every single day. Each day of our plan is different and has varying costs per serving per day. 

Some other ideas were tested to add to the variety of ingredients used, but all produced infeasible or not enough solutions. For example, we attempted to restrict the number of times each food item is allowed to repeat throughout the week, however, even if the allowed repeats were lifted to 4 times a week, it still produced infeasible solutions. We then tried to enforce this constraint on only vegetables and proteins, this only produced three different meal plans, and lifting constraints incrementally did not help us reaching our goal.

Eventually, we were able to arrive at our final model by constricting each sebsequent meal plan to have at least 8 different food items from the previous one, and this ensured that we have enough variety throughout the week. We wanted to obtain a more diverse meal plan than our current one, however, after tightening the constraint to 9 unique food items, the mdoel only produced 6 feasible meal plans.

This final updated model provides seven days of meal plan, while keeping the cost for all 7 days under 4 dollars per day.

The results are as follows: 

**Meal Plan 1:**

Total cost per serving ($): 2.66

Vegetables selections:
- Potatoes, Baked: 1.99 servings

Fruits selections:
-  Apple, Raw, w/Skin: 0.5 servings
-  Banana: 2.45 servings
-  Oranges: 3.0 servings

Protein selections:
-  Turkey: 0.5 servings
-  Pork: 0.5 servings

Grains selections:
-  Spaghetti W/ Sauce: 0.5 servings
-  White Bread: 0.5 servings

Dairy selections:
-  2pcnt Lowfat Milk: 2.80 servings
-  Skim Milk: 0.5 servings

**Meal Plan 2:**

Total cost per serving ($): 3.10

Vegetables selections:
-  Lettuce, Iceberg,Raw: 0.5 servings
-  Potatoes, Baked: 1.98 servings

Fruits selections:
-  Apple, Raw, w/Skin: 2.56 servings
-  Grapes: 0.5 servings
-  Oranges: 3.0 servings

Protein selections:
-  Beef: 0.5 servings
-  Pork: 0.5 servings

Grains selections:
-  Spaghetti W/ Sauce: 0.5 servings
-  Bagels: 0.5 servings
-  Wheat Bread: 0.5 servings

Dairy selections:
-  2pcnt Lowfat Milk: 2.79 servings
-  Skim Milk: 0.5 servings

**Meal Plan 3:**

Total cost per serving ($): 3.28

Vegetables selections:
-  Corn: 0.5 servings
-  Lettuce, Iceberg,Raw: 0.5 servings
-  Potatoes, Baked: 0.72 servings
-  Tomato,Red,Ripe,Raw: 0.5 servings

Fruits selections:
-  Apple, Raw, w/Skin: 2.59 servings
-  Banana: 1.72 servings
-  Oranges: 3.0 servings

Protein selections:
-  Scrambled Eggs: 0.5 servings
-  Pork: 0.5 servings

Grains selections:
-  Spaghetti W/ Sauce: 0.5 servings
-  Wheat Bread: 0.5 servings
-  Macaroni, cooked: 0.5 servings

Dairy selections:
-  2pcnt Lowfat Milk: 2.81 servings
-  Skim Milk: 0.5 servings

**Meal Plan 4:**

Total cost per serving ($): 3.45

Vegetables selections:
-  Lettuce, Iceberg,Raw: 0.5 servings
-  Peppers, Sweet, Raw: 0.5 servings
-  Potatoes, Baked: 1.84 servings

Fruits selections:
-  Apple, Raw, w/Skin: 2.94 servings
-  Kiwifruit, Raw, Fresh: 0.5 servings
-  Oranges: 3.0 servings

Protein selections:
-  Poached Eggs: 0.5 servings
-  Pork: 0.51 servings

Grains selections:
-  Spaghetti W/ Sauce: 0.55 servings
-  Macaroni, cooked: 0.5 servings

Dairy selections:
-  2pcnt Lowfat Milk: 2.75 servings
-  Skim Milk: 0.5 servings

**Meal Plan 5:**

Total cost per serving ($): 3.49

Vegetables selections:
-  Potatoes, Baked: 0.75 servings
-  Tomato,Red,Ripe,Raw: 0.5 servings

Fruits selections:
-  Apple, Raw, w/Skin: 3.0 servings
-  Banana: 3.0 servings
-  Grapes: 0.5 servings
-  Kiwifruit, Raw, Fresh: 0.60 servings

Protein selections:
-  Poached Eggs: 0.5 servings
-  Pork: 0.54 servings

Grains selections:
-  Spaghetti W/ Sauce: 0.5 servings
-  Wheat Bread: 0.5 servings

Dairy selections:
-  2pcnt Lowfat Milk: 3.0 servings
-  Skim Milk: 0.78 servings

**Meal Plan 6:**

Total cost per serving ($): 3.49

Vegetables selections:
-  Lettuce, Iceberg,Raw: 0.5 servings
-  Peppers, Sweet, Raw: 0.5 servings
-  Potatoes, Baked: 0.5 servings
-  Tomato,Red,Ripe,Raw: 0.5 servings

Fruits selections:
-  Apple, Raw, w/Skin: 2.36 servings
-  Banana: 2.20 servings
-  Kiwifruit, Raw, Fresh: 0.5 servings
-  Oranges: 3.0 servings

Protein selections:
-  Beef: 0.5 servings
-  Pork: 0.5 servings

Grains selections:
-  Wheat Bread: 1.06 servings
-  White Bread: 3.0 servings

Dairy selections:
-  2pcnt Lowfat Milk: 2.70 servings
-  Skim Milk: 0.5 servings

**Meal Plan 7:**

Total cost per serving ($): 3.52

Vegetables selections:
-  Corn: 0.5 servings
-  Potatoes, Baked: 1.63 servings
-  Tomato,Red,Ripe,Raw: 0.5 servings

Fruits selections:
-  Apple, Raw, w/Skin: 2.75 servings
-  Kiwifruit, Raw, Fresh: 0.5 servings
-  Oranges: 3.0 servings

Protein selections:
-  Beef: 0.5 servings
-  Pork: 0.5 servings

Grains selections:
-  Spaghetti W/ Sauce: 0.5 servings
-  Couscous: 0.5 servings

Dairy selections:
-  2pcnt Lowfat Milk: 2.81 servings
-  Skim Milk: 0.5 servings